velox/exec/TopNRowNumber.cpp

```diff
@@ -432,37 +422,34 @@ RowVectorPtr TopNRowNumber::getOutputFromMemory() {
   }
 
   vector_size_t offset = 0;
-  if (remainingRowsInPartition_ > 0) {
-    auto& partition = currentPartition();
-    auto start = partition.rows.size() - remainingRowsInPartition_;
-    const auto numRows =
-        std::min<vector_size_t>(outputBatchSize_, remainingRowsInPartition_);
-    appendPartitionRows(partition, start, numRows, offset, rowNumbers);
-    offset += numRows;
-    remainingRowsInPartition_ -= numRows;
-  }
-
+  // Continue to output as many remaining partitions as possible.
   while (offset < outputBatchSize_) {
-    auto* partition = nextPartition();
-    if (!partition) {
-      break;
+    // No previous partition to output (since this is the first partition).
+    if (!outputPartition_) {
+      outputPartition_ = nextPartition();
+      if (!outputPartition_) {
+        break;
+      }
     }
 
-    auto numRows = partition->rows.size();
-    if (offset + numRows > outputBatchSize_) {
-      remainingRowsInPartition_ = offset + numRows - outputBatchSize_;
-
-      // Add a subset of partition rows.
-      numRows -= remainingRowsInPartition_;
-      appendPartitionRows(*partition, 0, numRows, offset, rowNumbers);
-      offset += numRows;
+    auto numOutputRowsLeft = outputBatchSize_ - offset;
+    if (outputPartition_->rows.size() > numOutputRowsLeft) {
```

Comment:
If the current partition was already partially processed in the previous call, how do we know `remainingRowsInPartition_`?

Reply:
remainingRowsInPartition_ doesn't need to be computed as it is the size of the partition priority_queue at any time.

Reply:
What if the current partition was partially read in the previous call to getOutputFromMemory?
Do we remove the rows read from the current partition?

Reply:
Yes... they are being popped as they are being put in the outputRows_ buffer in appendPartitionRows